In [8]:
import sys
import subprocess

In [9]:
CHUNKS_PER_Z = 8064

In [10]:
def z_to_chunk(z):
    return z * CHUNKS_PER_Z

In [11]:
def resubmit(job_id, start_z, end_z):
    start_idx = z_to_chunk(start_z)
    stop_idx = z_to_chunk(end_z + 1)

    command = [
        'bsub',
        '-J', f'tiff_to_zarr3_s0_vol{job_id}',
        '-n', '25',
        '-W', '24:00',
        '-P', 'mengwang',
        '-g', '/scicompsoft/chend/tensorstore',
        '-o', f'/groups/scicompsoft/home/chend/temp/downsample_script/tensorswitch/output/output__vol{job_id}_%J.log',
        '-e', f'/groups/scicompsoft/home/chend/temp/downsample_script/tensorswitch/output/error_vol{job_id}_%J.log',
        '/groups/scicompsoft/home/chend/temp/downsample_script/tensorswitch/.pixi/envs/default/bin/python',
        '-m', 'tensorswitch',
        '--task', 'tiff_to_zarr3_s0',
        '--base_path', '/nrs/mengwang/jrc_celegans_lipl-4_Ariadne_Zarr/Dataset_B_lipl-4Tg/Dataset_B_nuclei_2/ulula2_nuclei',
        '--output_path', '/nrs/mengwang/jrc_celegans_lipl-4_Ariadne_Zarr/Dataset_B_lipl-4Tg/Dataset_B_nuclei_2/ulula2_nuclei/multiscale/s0',
        '--level', '0',
        '--downsample', '1',
        '--use_shard', '0',
        '--memory_limit', '50',
        '--start_idx', str(start_idx),
        '--stop_idx', str(stop_idx)
    ]

    print(f"Submitting Job {job_id}: z={start_z}–{end_z}, chunks={start_idx}–{stop_idx}")
    subprocess.run(command)

In [12]:
resubmit("1ba", 8, 9)

Submitting Job 1a: z=8–9, chunks=64512–80640
Job <145552151> is submitted to default queue <local>.


This job will be billed to mengwang


In [13]:
from string import ascii_lowercase

# Starting from z=10 to z=49, in steps of 5
z_start = 10
z_end = 49
chunk_size = 5

for i, z in enumerate(range(z_start, z_end + 1, chunk_size)):
    job_suffix = ascii_lowercase[i + 1]  # start from 'b', since 'a' was already used
    job_name = f"1{job_suffix}"
    start_z = z
    stop_z = min(z + chunk_size - 1, z_end)
    resubmit(job_name, start_z, stop_z)

Submitting Job 1b: z=10–14, chunks=80640–120960
Job <145552369> is submitted to default queue <local>.
Submitting Job 1c: z=15–19, chunks=120960–161280
Job <145552370> is submitted to default queue <local>.
Submitting Job 1d: z=20–24, chunks=161280–201600


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145552371> is submitted to default queue <local>.
Submitting Job 1e: z=25–29, chunks=201600–241920
Job <145552372> is submitted to default queue <local>.
Submitting Job 1f: z=30–34, chunks=241920–282240
Job <145552373> is submitted to default queue <local>.
Submitting Job 1g: z=35–39, chunks=282240–322560


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145552374> is submitted to default queue <local>.
Submitting Job 1h: z=40–44, chunks=322560–362880
Job <145552375> is submitted to default queue <local>.
Submitting Job 1i: z=45–49, chunks=362880–403200
Job <145552376> is submitted to default queue <local>.


This job will be billed to mengwang
This job will be billed to mengwang


In [15]:
from string import ascii_lowercase

z_start = 60
z_end = 100
chunk_size = 5

for i, z in enumerate(range(z_start, z_end + 1, chunk_size)):
    job_suffix = ascii_lowercase[i]  # start from 'a'
    job_name = f"2{job_suffix}"
    start_z = z
    stop_z = min(z + chunk_size - 1, z_end)
    resubmit(job_name, start_z, stop_z)

Submitting Job 2a: z=60–64, chunks=483840–524160
Job <145552927> is submitted to default queue <local>.
Submitting Job 2b: z=65–69, chunks=524160–564480
Job <145552928> is submitted to default queue <local>.
Submitting Job 2c: z=70–74, chunks=564480–604800


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145552929> is submitted to default queue <local>.
Submitting Job 2d: z=75–79, chunks=604800–645120
Job <145552930> is submitted to default queue <local>.
Submitting Job 2e: z=80–84, chunks=645120–685440
Job <145552931> is submitted to default queue <local>.
Submitting Job 2f: z=85–89, chunks=685440–725760


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145552932> is submitted to default queue <local>.
Submitting Job 2g: z=90–94, chunks=725760–766080
Job <145552933> is submitted to default queue <local>.
Submitting Job 2h: z=95–99, chunks=766080–806400
Job <145552934> is submitted to default queue <local>.
Submitting Job 2i: z=100–100, chunks=806400–814464
Job <145552935> is submitted to default queue <local>.


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


In [16]:
from string import ascii_lowercase

z_start = 112
z_end = 150
chunk_size = 5

for i, z in enumerate(range(z_start, z_end + 1, chunk_size)):
    job_suffix = ascii_lowercase[i]  # start from 'a'
    job_name = f"3{job_suffix}"
    start_z = z
    stop_z = min(z + chunk_size - 1, z_end)
    resubmit(job_name, start_z, stop_z)

Submitting Job 3a: z=112–116, chunks=903168–943488
Job <145553231> is submitted to default queue <local>.
Submitting Job 3b: z=117–121, chunks=943488–983808
Job <145553232> is submitted to default queue <local>.
Submitting Job 3c: z=122–126, chunks=983808–1024128


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145553233> is submitted to default queue <local>.
Submitting Job 3d: z=127–131, chunks=1024128–1064448
Job <145553234> is submitted to default queue <local>.
Submitting Job 3e: z=132–136, chunks=1064448–1104768
Job <145553235> is submitted to default queue <local>.
Submitting Job 3f: z=137–141, chunks=1104768–1145088


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145553236> is submitted to default queue <local>.
Submitting Job 3g: z=142–146, chunks=1145088–1185408
Job <145553237> is submitted to default queue <local>.
Submitting Job 3h: z=147–150, chunks=1185408–1217664
Job <145553238> is submitted to default queue <local>.


This job will be billed to mengwang
This job will be billed to mengwang


In [17]:
from string import ascii_lowercase

z_start = 162
z_end = 201
chunk_size = 5

for i, z in enumerate(range(z_start, z_end + 1, chunk_size)):
    job_suffix = ascii_lowercase[i]  # start from 'a'
    job_name = f"4{job_suffix}"
    start_z = z
    stop_z = min(z + chunk_size - 1, z_end)
    resubmit(job_name, start_z, stop_z)

Submitting Job 4a: z=162–166, chunks=1306368–1346688
Job <145553242> is submitted to default queue <local>.
Submitting Job 4b: z=167–171, chunks=1346688–1387008
Job <145553243> is submitted to default queue <local>.
Submitting Job 4c: z=172–176, chunks=1387008–1427328


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145553244> is submitted to default queue <local>.
Submitting Job 4d: z=177–181, chunks=1427328–1467648
Job <145553245> is submitted to default queue <local>.
Submitting Job 4e: z=182–186, chunks=1467648–1507968
Job <145553246> is submitted to default queue <local>.
Submitting Job 4f: z=187–191, chunks=1507968–1548288


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145553247> is submitted to default queue <local>.
Submitting Job 4g: z=192–196, chunks=1548288–1588608
Job <145553248> is submitted to default queue <local>.
Submitting Job 4h: z=197–201, chunks=1588608–1628928
Job <145553249> is submitted to default queue <local>.


This job will be billed to mengwang
This job will be billed to mengwang


In [18]:
from string import ascii_lowercase

z_start = 213
z_end = 252
chunk_size = 5

for i, z in enumerate(range(z_start, z_end + 1, chunk_size)):
    job_suffix = ascii_lowercase[i]  # start from 'a'
    job_name = f"5{job_suffix}"
    start_z = z
    stop_z = min(z + chunk_size - 1, z_end)
    resubmit(job_name, start_z, stop_z)

Submitting Job 5a: z=213–217, chunks=1717632–1757952
Job <145553253> is submitted to default queue <local>.
Submitting Job 5b: z=218–222, chunks=1757952–1798272
Job <145553254> is submitted to default queue <local>.
Submitting Job 5c: z=223–227, chunks=1798272–1838592


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145553255> is submitted to default queue <local>.
Submitting Job 5d: z=228–232, chunks=1838592–1878912
Job <145553256> is submitted to default queue <local>.
Submitting Job 5e: z=233–237, chunks=1878912–1919232
Job <145553257> is submitted to default queue <local>.
Submitting Job 5f: z=238–242, chunks=1919232–1959552


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145553258> is submitted to default queue <local>.
Submitting Job 5g: z=243–247, chunks=1959552–1999872
Job <145553259> is submitted to default queue <local>.
Submitting Job 5h: z=248–252, chunks=1999872–2040192
Job <145553260> is submitted to default queue <local>.


This job will be billed to mengwang
This job will be billed to mengwang


In [19]:
from string import ascii_lowercase

z_start = 264
z_end = 302
chunk_size = 5

for i, z in enumerate(range(z_start, z_end + 1, chunk_size)):
    job_suffix = ascii_lowercase[i]  # start from 'a'
    job_name = f"6{job_suffix}"
    start_z = z
    stop_z = min(z + chunk_size - 1, z_end)
    resubmit(job_name, start_z, stop_z)

Submitting Job 6a: z=264–268, chunks=2128896–2169216
Job <145553261> is submitted to default queue <local>.
Submitting Job 6b: z=269–273, chunks=2169216–2209536
Job <145553262> is submitted to default queue <local>.
Submitting Job 6c: z=274–278, chunks=2209536–2249856


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145553263> is submitted to default queue <local>.
Submitting Job 6d: z=279–283, chunks=2249856–2290176
Job <145553264> is submitted to default queue <local>.
Submitting Job 6e: z=284–288, chunks=2290176–2330496
Job <145553265> is submitted to default queue <local>.
Submitting Job 6f: z=289–293, chunks=2330496–2370816


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145553266> is submitted to default queue <local>.
Submitting Job 6g: z=294–298, chunks=2370816–2411136
Job <145553267> is submitted to default queue <local>.
Submitting Job 6h: z=299–302, chunks=2411136–2443392
Job <145553268> is submitted to default queue <local>.


This job will be billed to mengwang
This job will be billed to mengwang


In [20]:
from string import ascii_lowercase

z_start = 314
z_end = 353
chunk_size = 5

for i, z in enumerate(range(z_start, z_end + 1, chunk_size)):
    job_suffix = ascii_lowercase[i]  # start from 'a'
    job_name = f"7{job_suffix}"
    start_z = z
    stop_z = min(z + chunk_size - 1, z_end)
    resubmit(job_name, start_z, stop_z)

Submitting Job 7a: z=314–318, chunks=2532096–2572416
Job <145553269> is submitted to default queue <local>.
Submitting Job 7b: z=319–323, chunks=2572416–2612736
Job <145553270> is submitted to default queue <local>.
Submitting Job 7c: z=324–328, chunks=2612736–2653056


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145553271> is submitted to default queue <local>.
Submitting Job 7d: z=329–333, chunks=2653056–2693376
Job <145553272> is submitted to default queue <local>.
Submitting Job 7e: z=334–338, chunks=2693376–2733696
Job <145553273> is submitted to default queue <local>.
Submitting Job 7f: z=339–343, chunks=2733696–2774016


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145553274> is submitted to default queue <local>.
Submitting Job 7g: z=344–348, chunks=2774016–2814336
Job <145553275> is submitted to default queue <local>.
Submitting Job 7h: z=349–353, chunks=2814336–2854656
Job <145553276> is submitted to default queue <local>.


This job will be billed to mengwang
This job will be billed to mengwang


In [21]:
from string import ascii_lowercase

z_start = 365
z_end = 405
chunk_size = 5

for i, z in enumerate(range(z_start, z_end + 1, chunk_size)):
    job_suffix = ascii_lowercase[i]  # start from 'a'
    job_name = f"8{job_suffix}"
    start_z = z
    stop_z = min(z + chunk_size - 1, z_end)
    resubmit(job_name, start_z, stop_z)

Submitting Job 8a: z=365–369, chunks=2943360–2983680
Job <145553283> is submitted to default queue <local>.
Submitting Job 8b: z=370–374, chunks=2983680–3024000
Job <145553284> is submitted to default queue <local>.
Submitting Job 8c: z=375–379, chunks=3024000–3064320


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145553285> is submitted to default queue <local>.
Submitting Job 8d: z=380–384, chunks=3064320–3104640
Job <145553286> is submitted to default queue <local>.
Submitting Job 8e: z=385–389, chunks=3104640–3144960
Job <145553287> is submitted to default queue <local>.
Submitting Job 8f: z=390–394, chunks=3144960–3185280


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


Job <145553288> is submitted to default queue <local>.
Submitting Job 8g: z=395–399, chunks=3185280–3225600
Job <145553289> is submitted to default queue <local>.
Submitting Job 8h: z=400–404, chunks=3225600–3265920
Job <145553290> is submitted to default queue <local>.
Submitting Job 8i: z=405–405, chunks=3265920–3273984
Job <145553291> is submitted to default queue <local>.


This job will be billed to mengwang
This job will be billed to mengwang
This job will be billed to mengwang


0-8: within 8 up to 58
50-60: within 60 up to 44
101-112: within 112 up to 26
151-162: within 162 up to 83
202-213: within 213 up to 51
253-264: within 264 up to 11
303-314: within 314 up to 75
354-365: within 365 up to 35

| Z-slice | Start Chunk Index                |
| ------- | -------------------------------- |
| 8       | 64,512                           |
| 60      | 483,840                          |
| 112     | 903,168                          |
| 162     | 1,306,368                        |
| 213     | 1,717,632                        |
| 264     | 2,128,896                        |
| 314     | 2,532,096                        |
| 365     | 2,943,360                        |

(tensorswitch) [jupyterhub - chend@e06u07]~/temp/downsample_script/tensorswitch>python z_to_chunk_index.py /nrs/mengwang/jrc_celegans_lipl-4_Ariadne_Zarr/Dataset_B_lipl-4Tg/Dataset_B_nuclei_2/ulula2_nuclei 8
Volume shape: (z=25860, y=6124, x=5328)
Chunks per z-slice: 8064
Z-slice 8 → Start chunk index: 64512
(tensorswitch) [jupyterhub - chend@e06u07]~/temp/downsample_script/tensorswitch>python z_to_chunk_index.py /nrs/mengwang/jrc_celegans_lipl-4_Ariadne_Zarr/Dataset_B_lipl-4Tg/Dataset_B_nuclei_2/ulula2_nuclei 60
Volume shape: (z=25860, y=6124, x=5328)
Chunks per z-slice: 8064
Z-slice 60 → Start chunk index: 483840
(tensorswitch) [jupyterhub - chend@e06u07]~/temp/downsample_script/tensorswitch>python z_to_chunk_index.py /nrs/mengwang/jrc_celegans_lipl-4_Ariadne_Zarr/Dataset_B_lipl-4Tg/Dataset_B_nuclei_2/ulula2_nuclei 112
Volume shape: (z=25860, y=6124, x=5328)
Chunks per z-slice: 8064
Z-slice 112 → Start chunk index: 903168
(tensorswitch) [jupyterhub - chend@e06u07]~/temp/downsample_script/tensorswitch>python z_to_chunk_index.py /nrs/mengwang/jrc_celegans_lipl-4_Ariadne_Zarr/Dataset_B_lipl-4Tg/Dataset_B_nuclei_2/ulula2_nuclei 162
Volume shape: (z=25860, y=6124, x=5328)
Chunks per z-slice: 8064
Z-slice 162 → Start chunk index: 1306368
(tensorswitch) [jupyterhub - chend@e06u07]~/temp/downsample_script/tensorswitch>python z_to_chunk_index.py /nrs/mengwang/jrc_celegans_lipl-4_Ariadne_Zarr/Dataset_B_lipl-4Tg/Dataset_B_nuclei_2/ulula2_nuclei 213
Volume shape: (z=25860, y=6124, x=5328)
Chunks per z-slice: 8064
Z-slice 213 → Start chunk index: 1717632
(tensorswitch) [jupyterhub - chend@e06u07]~/temp/downsample_script/tensorswitch>python z_to_chunk_index.py /nrs/mengwang/jrc_celegans_lipl-4_Ariadne_Zarr/Dataset_B_lipl-4Tg/Dataset_B_nuclei_2/ulula2_nuclei 264
Volume shape: (z=25860, y=6124, x=5328)
Chunks per z-slice: 8064
Z-slice 264 → Start chunk index: 2128896
(tensorswitch) [jupyterhub - chend@e06u07]~/temp/downsample_script/tensorswitch>python z_to_chunk_index.py /nrs/mengwang/jrc_celegans_lipl-4_Ariadne_Zarr/Dataset_B_lipl-4Tg/Dataset_B_nuclei_2/ulula2_nuclei 314
Volume shape: (z=25860, y=6124, x=5328)
Chunks per z-slice: 8064
Z-slice 314 → Start chunk index: 2532096
(tensorswitch) [jupyterhub - chend@e06u07]~/temp/downsample_script/tensorswitch>python z_to_chunk_index.py /nrs/mengwang/jrc_celegans_lipl-4_Ariadne_Zarr/Dataset_B_lipl-4Tg/Dataset_B_nuclei_2/ulula2_nuclei 365
Volume shape: (z=25860, y=6124, x=5328)
Chunks per z-slice: 8064
Z-slice 365 → Start chunk index: 2943360

In [6]:
def resubmit(job_number, start_idx, stop_idx):
    command = ['bsub',
     '-J',
     f'tiff_to_zarr3_s0_vol{job_number}',
     '-n',
     '24',
     '-W',
     '24:00',
     '-P',
     'mengwang',
     '-g',
     '/scicompsoft/chend/tensorstore',
     '-o',
     f'/groups/scicompsoft/home/chend/temp/downsample_script/tensorswitch/output/output__vol{job_number}_%J.log',
     '-e',
     f'/groups/scicompsoft/home/chend/temp/downsample_script/tensorswitch/output/error_vol{job_number}_%J.log',
     '/groups/scicompsoft/home/chend/temp/downsample_script/tensorswitch/.pixi/envs/default/bin/python',
     '-m',
     'tensorswitch',
     '--task',
     'tiff_to_zarr3_s0',
     '--base_path',
     '/nrs/mengwang/jrc_celegans_lipl-4_Ariadne_Zarr/Dataset_B_lipl-4Tg/Dataset_B_nuclei_2/ulula2_nuclei',
     '--output_path',
     '/nrs/mengwang/jrc_celegans_lipl-4_Ariadne_Zarr/Dataset_B_lipl-4Tg/Dataset_B_nuclei_2/ulula2_nuclei/multiscale/s0',
     '--level',
     '0',
     '--downsample',
     '1',
     '--use_shard',
     '0',
     '--memory_limit',
     '50',
     '--start_idx',
     str(start_idx),
     '--stop_idx',
     str(stop_idx)
    ]
    print(command)
    subprocess.run(command)

In [ ]:
resubmit(2a,483840,516096)

In [ ]:
64512 - 385452